sound/soc/sof/intel/hda-ipc.c

```diff
@@ -67,6 +67,21 @@ int hda_dsp_ipc_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
 	return 0;
 }
 
+void hda_dsp_ipc4_schedule_d0i3_work(struct sof_intel_hda_dev *hdev,
+				     struct snd_sof_ipc_msg *msg)
+{
+	struct sof_ipc4_msg *msg_data = msg->msg_data;
+
+	/* Schedule a delayed work for d0i3 entry after sending non-pm gate ipc msg */
+	if (SOF_IPC4_MSG_IS_MODULE_MSG(msg_data->primary) &&
+	    (SOF_IPC4_MSG_TYPE_GET(msg_data->primary) == SOF_IPC4_MOD_SET_DX ||
+	    SOF_IPC4_MSG_TYPE_GET(msg_data->primary) == SOF_IPC4_MOD_SET_D0IX))
```

Comment:
it would be more readable to use a helper or static inline that deals with all the details

Reply:
sure, thanks